learning-from-errors/overlapping-instances/article.md

```diff
@@ -0,0 +1,626 @@
+
+```hs
+
+{-# LANGUAGE FlexibleInstances #-}
+
+module Main (main) where
+
+class Printable a where
+  printMe :: a -> IO ()
+
+instance Printable a where
+  printMe a = putStrLn "general instance"
+
+instance {-# INCOHERENT #-} Printable Int where
+  printMe a = putStrLn "int instance"
+
+main :: IO ()
+main = fn (5 :: Int)
+
+fn :: a -> IO ()
+fn x = printMe x
+```
+</details>
+
+The rules followed by the instance resolution algorithm are described
+[here](https://ghc.gitlab.haskell.org/ghc/doc/users_guide/exts/instances.html#extension-IncoherentInstances)
+and in this specific case, the general instance will get applied because it
+ends up being the single 'prime candidate' which gets selected since the
+remaining instance is marked as `INCOHERENT`. This means that the program
+will print "general instance" if you run it.
+
+But what if you want the `Int` instance to be called instead? One way is to use
+`Typeable` and prove to the compiler that `a` is, in fact, an `Int`. And once
+you prove that `a` is an `Int`, GHC will happily call the `printMe` in the `Int`
+instance for you. Of course, this will happen at the cost of including a `Typeable` constraint in
+the signature of the `fn` function (and a small runtime cost added to the function because of the Typeable
+constraint).
+
+```hs
+fn :: forall a. Typeable a => a -> IO ()
+fn x = case cast x of
+  (Just a :: Maybe Int) -> printMe a
+  Nothing               -> printMe x
+```
+
+<details>
+  <summary>Full code</summary>
+
+```hs
+{-# LANGUAGE FlexibleInstances   #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Main (main) where
+
+import Data.Typeable (cast)
+
+class Printable a where
+  printMe :: a -> IO ()
+
+instance Printable a where
+  printMe a = putStrLn "general instance"
+
+instance {-# INCOHERENT #-} Printable Int where
+  printMe a = putStrLn "int instance"
+
+main :: IO ()
+main = fn (5 :: Int)
+
+fn :: forall a. Typeable a => a -> IO ()
+fn x = case cast x of
+  (Just a :: Maybe Int) -> printMe a
+  Nothing               -> printMe x
+```
+</details>
+
+Similarly, we could change/hack the general instance to use `Typeable` and implement
+specialized behavior for `Int` in the general instance itself.
+
+```hs
+instance Typeable a => Printable a where
+  printMe a = case cast a of
+    (Just x :: Maybe Int) -> putStrLn "general instance for int"
+    Nothing               -> putStrLn "general instance"
+```
+
+<details>
+  <summary>Full code</summary>
+
+```hs
+{-# LANGUAGE FlexibleInstances   #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Main (main) where
+
+import Data.Typeable
+
+class Printable a where
+  printMe :: a -> IO ()
+
+instance Typeable a => Printable a where
+  printMe a = case cast a of
+    (Just x :: Maybe Int) -> putStrLn "general instance for int"
+    Nothing               -> putStrLn "general instance"
+
+main :: IO ()
+main = fn (5 :: Int)
+
+fn :: forall a. Typeable a => a -> IO ()
+fn x = printMe x
+```
+</details>
```

Comment:
> why GHC is not picking up the general instance to use in `fn`

Well, it’s not clear that the user would want this behavior. In fact, it would be good to explicitly spell it out in the article.

Say that there are two reasonable behaviors here:
1.  Just use the default instance (then we need INCOHERENT)
2. Postpone the decision to the use site of `fn` (then we need a either `Printable a` or `Typeable a`)